src/dailyai/pipeline/pipeline.py

```diff
@@ -96,18 +101,32 @@ async def run_pipeline(self):
             # here.
             for processor in self._processors:
                 await processor.interrupted()
-            pass
 
     async def _run_pipeline_recursively(
-        self, initial_frame: Frame, processors: List[FrameProcessor]
+        self, initial_frame: Frame, processors: List[FrameProcessor], depth=1
     ) -> AsyncGenerator[Frame, None]:
         """Internal function to add frames to the pipeline as they're yielded
         by each processor."""
         if processors:
+            self._log_frame(initial_frame, depth)
             async for frame in processors[0].process_frame(initial_frame):
                 async for final_frame in self._run_pipeline_recursively(
-                    frame, processors[1:]
+                    frame, processors[1:], depth + 1
                 ):
                     yield final_frame
         else:
             yield initial_frame
+
+    def _log_frame(self, frame: Frame, depth: int):
+        if not isinstance(frame, AudioFrame):
+            if depth == 0:
+                source = "source"
+                dest = str(self._processors[depth - 1])
+            elif depth == len(self._processors) + 1:
+                dest = "sink"
+                source = str(self._processors[depth - 2])
+            else:
+                dest = str(self._processors[depth - 1])
+                source = str(self._processors[depth - 2])
```

Comment:
same here. if `depth` is 1 will the index be -1?

Reply:
Argh, you're right. Fixed and added tests.